src/overlays/actors/ovl_Door_Spiral/z_door_spiral.c

```diff
@@ -31,30 +40,223 @@ const ActorInit Door_Spiral_InitVars = {
+    { OBJECT_IKANA_OBJ, 5 }, { OBJECT_DANPEI_OBJECT, 7 },  { OBJECT_IKNINSIDE_OBJ, 6 },
 };
-*/
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Door_Spiral/func_809A2B60.s")
+typedef struct {
+    /* 0x0 */ s16 sceneId;
+    /* 0x2 */ u8 index;
+} SpiralSceneInfo; // size = 0x4
+
+SpiralSceneInfo sSpiralSceneInfoTable[] = {
+    { SCENE_MITURIN, 2 },     { SCENE_HAKUGIN, 3 },   { SCENE_INISIE_N, 4 }, { SCENE_INISIE_R, 4 },
+    { SCENE_DANPEI2TEST, 5 }, { SCENE_IKNINSIDE, 6 }, { SCENE_CASTLE, 6 },
+};
+
+static InitChainEntry sInitChain[] = {
+    ICHAIN_VEC3F(scale, 1, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneForward, 4000, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneScale, 400, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneDownward, 400, ICHAIN_STOP),
+};
+
+void DoorSpiral_SetupAction(DoorSpiral* this, DoorSpiralActionFunc actionFunc) {
+    this->actionFunc = actionFunc;
+    this->unk14A = 0;
+}
+
+s32 func_809A2B70(DoorSpiral* this, PlayState* play) {
+    SpiralObjectInfo* spiralObjectInfo = &sSpiralObjectInfoTable[this->unk147];
+
+    this->unk148 = spiralObjectInfo->index;
+    if ((this->unk148 == 7) || ((this->unk148 == 2) && play->roomCtx.curRoom.enablePosLights)) {
+        if (this->unk148 == 2) {
+            this->unk148 = 3;
+        }
+        this->actor.flags |= ACTOR_FLAG_10000000;
+    }
+    DoorSpiral_SetupAction(this, func_809A2FF8);
+    return 0;
+}
+
+u8 func_809A2BF8(PlayState* play) {
+    SpiralSceneInfo* spiralSceneInfo = sSpiralSceneInfoTable;
+    s32 i;
+    u8 var_v1;
+
+    for (i = 0; i < ARRAY_COUNT(sSpiralSceneInfoTable); i++) {
+        if (play->sceneId == spiralSceneInfo->sceneId) {
+            break;
+        }
+        spiralSceneInfo++;
+    }
+    if (i < ARRAY_COUNT(sSpiralSceneInfoTable)) {
+        var_v1 = spiralSceneInfo->index;
+    } else {
+        var_v1 = (Object_GetSlot(&play->objectCtx, GAMEPLAY_DANGEON_KEEP) > OBJECT_SLOT_NONE) ? 1 : 0;
+    }
+    return var_v1;
+}
+
+void DoorSpiral_Init(Actor* thisx, PlayState* play) {
+    DoorSpiral* this = THIS;
+    s32 transitionActorId = DOORSPIRAL_GET_FC00(thisx);
+    s8 objectSlot;
+
+    if (this->actor.room != play->doorCtx.transitionActorList[transitionActorId].sides[0].room) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    this->type = DOORSPIRAL_GET_TYPE(thisx);
+    this->direction = DOORSPIRAL_GET_80(thisx);
+    this->unk147 = func_809A2BF8(play);
+    objectSlot = Object_GetSlot(&play->objectCtx, sSpiralObjectInfoTable[this->unk147].objectId);
+    this->objSlot = objectSlot;
+    if (objectSlot < 0) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+    DoorSpiral_SetupAction(this, func_809A2DB0);
+    Actor_SetFocus(&this->actor, 60.0f);
+}
+
+void DoorSpiral_Destroy(Actor* thisx, PlayState* play) {
+    s32 transitionActorId = DOORSPIRAL_GET_FC00(thisx);
```

Comment:
```suggestion
    s32 transitionId = DOOR_GET_TRANSITION_ID(thisx);
```
I might get others thoughts on this, but should this use the common `DOOR_GET_TRANSITION_ID` param getter, or should the spiral door have it's own macro?

Also, having the prefix be `ActorId` makes me think it takes the `ActorId` enum, so just `transitionId` should be fine

Reply:
This is a more open-ended question so I wouldn't be quick to resolve this. I'd still like others thoughts on this if possible

Reply:
As far as I can tell all doors should use this macro. 

Player uses `DOOR_GET_TRANSITION_ID` for the staircase door
https://github.com/zeldaret/mm/blob/2c182bf7e35f6b7ac7853071875343f591c37510/src/overlays/actors/ovl_player_actor/z_player.c#L6329-L6330

Reply:
I suppose in theory you could use `DOORSPIRAL_GET_TRANSITION_ID` in this player function `Player_Door_Staircase`, but it makes sense to stick with a more consistent `DOOR_GET_TRANSITION_ID`